multik-jvm/src/main/kotlin/org/jetbrains/kotlinx/multik/jvm/JvmLinAlg.kt

```diff
@@ -20,29 +29,222 @@ public object JvmLinAlg : LinAlg {
+     */
+    public fun solveTriangle(a: D2Array<Double>, b: D2Array<Double>, isLowerTriangle: Boolean = true): D2Array<Double> {
+        require(a.shape[1] == b.shape[0]) { "invalid arguments, a.shape[1] = ${a.shape[1]} != b.shape[0]=${b.shape[0]}" }
+        require(a.shape[0] == a.shape[1]) { "a should be a square matrix, matrix with shape (${a.shape[0]}, ${a.shape[1]}) given" }
+        val x = b.deepCopy()
+        for (i in 0 until x.shape[0]) {
+            for (j in 0 until x.shape[1]) {
+                x[i, j] /= a[i, i]
+            }
+        }
+        if (isLowerTriangle) {
+            for (i in 0 until x.shape[0]) {
+                for (k in i + 1 until x.shape[0]) {
+                    for (j in 0 until x.shape[1]) {
+                        x[k, j] -= a[k, i] * x[i, j] / a[k, k]
+                    }
+                }
+            }
+        } else {
+            for (i in x.shape[0] - 1 downTo 0) {
+                for (k in i - 1 downTo 0) {
+                    for (j in 0 until x.shape[1]) {
+                        x[k, j] -= a[k, i] * x[i, j] / a[k, k]
+                    }
+                }
+            }
+
+        }
+        return x
+    }
+
+    fun solveDouble(a: D2Array<Double>, b: D2Array<Double>, singularityErrorLevel: Double = 1e-7): D2Array<Double> {
+        require(a.shape[0] == a.shape[1] && a.shape[1] == b.shape[0])
+        val (P, L, U) = PLUCompressed(a)
+        val _b = b.deepCopy()
+
+        for (i in P.indices) {
+            if(P[i] != 0) {
+                _b[i] = _b[i + P[i]].deepCopy().also { _b[i + P[i]] = _b[i].deepCopy() }
+            }
+        }
+        for (i in 0 until U.shape[0]) {
+            if (abs(U[i, i]) < singularityErrorLevel) {
+                throw ArithmeticException("matrix a is singular or almost singular")
+            }
+        }
+
+        return solveTriangle(U, solveTriangle(L, _b), false)
+    }
+
+    override fun <T : Number, D : Dim2> solve(a: MultiArray<T, D2>, b: MultiArray<T, D>): NDArray<T, D> {
+        if(a.dtype != DataType.DoubleDataType) {
+            throw UnsupportedOperationException()
+        }
+        val aDouble = mk.d2array<Double>(a.shape[0], a.shape[1]) { 0.0 }
+        for (i in 0 until a.shape[0]) {
+            for (j in 0 until a.shape[1]) {
+                aDouble[i, j] = a[i, j].toDouble()
+            }
+        }
+        val bDouble: D2Array<Double>
+        if (b.dim.d == 2) {
+            b as D2Array<Double>
+            bDouble = mk.d2array(b.shape[0], b.shape[1]) { 0.0 }
+            for (i in 0 until b.shape[0]) {
+                for (j in 0 until b.shape[1]) {
+                    bDouble[i, j] = b[i, j].toDouble()
+                }
+            }
+        } else {
+            b as D1Array<Double>
+            bDouble = mk.d2array(b.shape[0], 1) { 0.0 }
+            for (i in 0 until b.shape[0]) {
+                bDouble[i, 0] = b[i]
+            }
+
+        }
```

Comment:
Similar to above

Reply:
Agree with the first case, but second case isn't similar. I want to cast D1 array to D2 array (add new axis), can you please help what is the best way to do it?